File: node-pag/examples/file.js
const path = require('path');

const {
  PAGFile,
  PAGImage,
  LayerType,
} = require('..');

// static methods
console.log("MaxSupportedTagLevel: ", PAGFile.MaxSupportedTagLevel());

// image
const image = new PAGImage(path.join(__dirname, '../../third_party/libpag/assets/mountain.jpg'));
console.log('>>>> image.unique_id: ', image.unique_id);
console.log('width: ', image.width);
console.log('height: ', image.height);
console.log('scale_mode: ', image.scale_mode);

// file
const file = new PAGFile(path.join(__dirname, '../../third_party/libpag/assets/AudioMarker.pag'));

// PAGLayer methods
console.log('>>>> file.unique_id: ', file.unique_id);
console.log('width: ', file.width);
console.log('height: ', file.height);
console.log('duration: ', file.duration);
console.log('frame_rate: ', file.frame_rate);
console.log('start_time: ', file.start_time);
console.log('current_time: ', file.current_time);
console.log('progress: ', file.progress);
console.log('tag_level: ', file.tag_level);
console.log('layer_name: ', file.layer_name);
console.log('layer_type: ', file.layer_type);
console.log('alpha: ', file.alpha);
console.log('visible: ', file.visible);

// PAGComposition methods
console.log('num_children: ', file.num_children);
console.log('audio_start_time: ', file.audio_start_time);

// PAGFile methods
console.log('num_texts: ', file.num_texts);
console.log('num_images: ', file.num_images);
console.log('num_videos: ', file.num_videos);
console.log('path: ', file.path);
console.log('time_stretch_mode', file.time_stretch_mode);
const text_indices = file.get_editable_indices(LayerType.Text);
console.log("file editable text indices", text_indices);
const image_indices = file.get_editable_indices(LayerType.Image);
console.log("file editable image indices", image_indices);

File: examples/src/bin/file.rs
use libpag::*;
use std::path::PathBuf;

fn main() {
    let dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let assets_dir = dir.join("../third_party/libpag/assets");

    let invalid_file = PAGFile::from_file(assets_dir.join("not-a-pag-file.pag").to_str().unwrap());
    println!("invalid_file is {:?}", invalid_file);

    let pag_file = PAGFile::from_file(assets_dir.join("MT2.pag").to_str().unwrap()).unwrap();
    println!("tag level = {}", pag_file.tag_level());

    println!("text num = {}", pag_file.num_texts());
    println!("image num = {}", pag_file.num_images());
    println!("video num = {}", pag_file.num_videos());

    println!("path = {}", pag_file.path());
    println!("time_stretch_mode = {:?}", pag_file.time_stretch_mode());
    println!("is_pag_file = {}", pag_file.is_pag_file());

    // 父类 PAGComposition 方法
    println!("num_children = {}", pag_file.num_children());
    println!("width = {}", pag_file.width());
    println!("width = {}", pag_file.height());
    println!("audio_bytes = {:?}", pag_file.audio_bytes());

    // 父类 PAGLayer 方法
    println!("unique_id = {}", pag_file.unique_id());
    println!("layer_type = {}", pag_file.layer_type());
    println!("layer_name = {}", pag_file.layer_name());
    println!("alpha = {}", pag_file.alpha());
    pag_file.set_alpha(0.4);
    println!("alpha = {}", pag_file.alpha());
    println!("get_bounds = {}", pag_file.get_bounds());
    println!("visible = {}", pag_file.visible());
    pag_file.set_visible(false);
    println!("visible = {}", pag_file.visible());

    println!("editable_index = {}", pag_file.editable_index());
    println!("parent is None = {}", pag_file.parent().is_none());
    println!(
        "track_matte_layer is None = {}",
        pag_file.track_matte_layer().is_none()
    );

    println!("duration = {}", pag_file.duration());
    println!("frame_rate = {}", pag_file.frame_rate());
    println!("start_time = {}", pag_file.start_time());
    println!("current_time = {}", pag_file.current_time());
    println!("progress = {}", pag_file.get_progress());
    pag_file.next_frame();
    println!("start_time = {}", pag_file.start_time());
    println!("current_time = {}", pag_file.current_time());
    println!("progress = {}", pag_file.get_progress());
    pag_file.set_start_time(10);
    println!("start_time = {}", pag_file.start_time());
    println!("current_time = {}", pag_file.current_time());
    println!("progress = {}", pag_file.get_progress());

    println!(
        "excluded_from_timeline = {}",
        pag_file.excluded_from_timeline()
    );
    pag_file.set_excluded_from_timeline(true);
    println!(
        "excluded_from_timeline = {}",
        pag_file.excluded_from_timeline()
    );

    let matrix = pag_file.matrix();
    println!("matrix: {:?}", matrix.isIdentity());
}
